src/components/channels/ChannelsSubscriber.js

```diff
@@ -23,99 +23,117 @@ const VIEWER_PALETTE = [
 
-  componentWillUnmount() {
-    PubSub.unsubscribe(this.rasterAddToken);
-  }
+      const initialColorValues = domains.map((_, i) => VIEWER_PALETTE[i]);
+      setColorValues(initialColorValues);
+      PubSub.publish(COLORS_CHANGE + id, initialColorValues);
 
-  rasterAddSubscriber(msg, rasterData) {
-    Object.keys(rasterData)
-      .forEach((channel, i) => {
-        const rangeValue = { [channel]: rasterData[channel].range };
-        this.setState(prevState => ({ rangeValues: { ...prevState.rangeValues, ...rangeValue } }));
-        this.setColorValue({ channel, rgb: VIEWER_PALETTE[i] });
-        this.toggleChannel(channel);
+      const initialSliderValues = domains.map((d) => {
+        const isArray = Array.isArray(d);
+        return isArray ? [d[0], Math.ceil(d[1] / 5)] : [0, Math.ceil(STANDARD_MAX / 5)];
       });
-  }
+      // "5" is arbitrary, but the data tends to be left-skewed.
+      // Eventually we want this to be based on the data in the image.));
+      setSliderValues(initialSliderValues);
+      PubSub.publish(SLIDERS_CHANGE + id, initialSliderValues);
 
-  // eslint-disable-next-line class-methods-use-this
-  setSliderValue(sliderValue) {
-    PubSub.publish(SLIDERS_CHANGE, sliderValue);
-  }
+      const initialChannelIsOn = Array(domains.length).fill(true);
+      setChannelIsOn(initialChannelIsOn);
+      PubSub.publish(CHANNEL_TOGGLE + id, initialChannelIsOn);
 
-  setColorValue({ channel, rgb }) {
-    const colorValue = { [channel]: rgb };
-    this.setState(prevState => ({ colorValues: { ...prevState.colorValues, ...colorValue } }));
-    PubSub.publish(COLORS_CHANGE, { channel, rgb });
-  }
 
-  toggleChannel(channel) {
-    this.setState((prevState) => {
-      const channelToggle = !prevState.channelsOn[channel];
-      const channelOn = { [channel]: channelToggle };
-      PubSub.publish(CHANNEL_TOGGLE, { channel, channelToggle });
-      return { channelsOn: { ...prevState.channelsOn, ...channelOn } };
+      setChannelNames(cNames);
+      setSliderDomains(domains.map(d => (Array.isArray(d) ? d : [0, STANDARD_MAX])));
+      setEventId(id);
+    }
+    memoizedOnReady();
+    const token = PubSub.subscribe(RASTER_ADD, handleRasterAdd);
+    return () => PubSub.unsubscribe(token);
+  }, [memoizedOnReady]);
+
+  const handleColorChange = (i, rgb) => {
+    setColorValues((prevColorValues) => {
+      const nextColorValues = [...prevColorValues];
+      nextColorValues[i] = rgb;
+      PubSub.publish(COLORS_CHANGE + eventId, nextColorValues);
+      return nextColorValues;
     });
-  }
+  };
+
+  const handleSliderChange = (i, sliderValue) => {
+    setSliderValues((prevSliderValues) => {
+      const nextSliderValues = [...prevSliderValues];
+      nextSliderValues[i] = sliderValue;
+      PubSub.publish(SLIDERS_CHANGE + eventId, nextSliderValues);
+      return nextSliderValues;
+    });
+  };
+
+  const handleChannelIsOnChange = (i) => {
+    setChannelIsOn((prevChannelIsOn) => {
```

Comment:
This doesn't really set the channel state: it toggles it. Probably ok, but I think it might be good to make it a setter like the others, primarily for consistency... but if the browser was really lagging, you could imagine:
- Channel is on.
- User clicks to toggle off.
- (... browser is busy and doesn't respond to first click ...)
- Impatient user clicks again.
- both events are finally processed... and the channel is still on.

or maybe my expectation on a double click is to get back to the original state, even if it takes a while? I'm not sure: may depend on how snappy we expect the interface to be?

Reply:
Things feel pretty snappy since this is being delt within the setChannelIsOn callback, but I can explore more. I'm not sure if I've experienced the described issue. At minimum, I'll change `setChannelIsOn` to `toggleChannelIsOn`

Reply:
Let's leave it as is; sorry to bring it up.